chore: allow cloudfront distribution creation via env manifest

CaptainCarpensir opened the pull request:
<!-- Provide summary of changes -->
Allows CloudFront to be deployed in front of a load balancer with no certificate/domain set. A following PR will be made to allow the user to configure the environment ALB to allow public traffic to the load balancer or not while CDN is enabled.

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->
PR as part of design in #3701 which satisfies requirements 1, 2, 4, 6, and 12. Milestone 1 still not met with this PR.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

Commit:
chore(manifest): add cdn bool to env manifest

Commit:
chore(manifest): add cdn bool to env manifest

CaptainCarpensir committed:
Merge branch 'aws:mainline' into env-manifest-cdn

Commit:
allow cfn to create barebones cf dist via stack

Commit:
Update CDN stack template

Commit:
add test, change cf public lb ingress

Commit:
Merge branch 'mainline' into env-manifest-cdn

Commit:
get prefix list for public lb security group

Commit:
correct ingress logic

CaptainCarpensir committed:
Merge branch 'mainline' into env-manifest-cdn

Commit:
rename manifest tag to be more clear

Commit:
Merge branch 'env-manifest-cdn' of github.com:CaptainCarpensir/copilot-cli into env-manifest-cdn

iamhopaul123 commented on internal/pkg/cli/deploy/env.go:
```diff
@@ -104,6 +111,24 @@ func (d *envDeployer) UploadArtifacts() (map[string]string, error) {
 	return d.legacyUploadCustomResources(resources.S3Bucket)
 }
 
+func (d *envDeployer) getPrefixListId(in *DeployEnvironmentInput) (*string, error) {
```
```suggestion
func (d *envDeployer) cfManagedPrefixListID(in *DeployEnvironmentInput) (*string, error) {
```

iamhopaul123 commented on internal/pkg/cli/deploy/env.go:
```diff
@@ -104,6 +111,24 @@ func (d *envDeployer) UploadArtifacts() (map[string]string, error) {
+
+	if d.ec2 == nil {
+		return nil, nil
+	}
```
We don't have to do this. we just need to make sure we always use `NewEnvDeployer` to initialize the struct.

iamhopaul123 commented on internal/pkg/cli/deploy/env.go:
```diff
@@ -104,6 +111,24 @@ func (d *envDeployer) UploadArtifacts() (map[string]string, error) {
 }
 
+func (d *envDeployer) getPrefixListId(in *DeployEnvironmentInput) (*string, error) {
+	if in.Manifest != nil {
```
We probably don't need this validation because the manifest should be there all the time (@Lou1415926 is it true?). Otherwise we might update the stack and unexpectedly remove the ingress control because of that.

iamhopaul123 commented on internal/pkg/template/env.go:
```diff
@@ -103,10 +104,17 @@ type EnvOpts struct {
 	SerializedManifest string // Serialized manifest used to render the environment template.
 }
 
+type CDNConfig struct {
```
Comment pls

Commit:
fix unit test manifest example

iamhopaul123 commented on internal/pkg/template/env.go:
```diff
@@ -103,10 +104,17 @@ type EnvOpts struct {
 	Telemetry                *Telemetry
 
-	LatestVersion      string
+	CDNConfig CDNConfig
```
```suggestion
	CDNConfig *CDNConfig
```
Then maybe we don't need `EnableCDN bool` at all?

iamhopaul123 commented on internal/pkg/template/templates/environment/partials/cdn-resources.yml:
```diff
@@ -0,0 +1,24 @@
+            OriginProtocolPolicy: http-only # Will need to change with TLS termination and Alias usage
+          DomainName: !GetAtt PublicLoadBalancer.DNSName
+          Id: !Sub 'copilot-${AppName}-${EnvironmentName}-origin'
+      # ViewerCertificate: # Used for Aliases
+      #   AcmCertificateArn: # Imported certificate goes here
+      #   CloudFrontDefaultCertificate: true # If importing certificate, set to false
+      #   MinimumProtocolVersion: TLSv1 # Default for now, may need to change later
+      #   SsLSupportMethod: sni-only # Changing this can incur costs to the user
```
do we need to get rid of these?

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/env.go:
```diff
@@ -296,6 +297,33 @@ func (e *BootstrapEnvStackConfig) ToEnv(stack *cloudformation.Stack) (*config.En
 	}, nil
 }
 
+func (e *EnvStackConfig) cdnConfig() template.CDNConfig {
```
nit we could return `*template.CDNConfig` to simplify the logic in this function and use

```
{{- if .CDNConfig}}
```

in the template file.

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/env.go:
```diff
@@ -296,6 +297,33 @@ func (e *BootstrapEnvStackConfig) ToEnv(stack *cloudformation.Stack) (*config.En
+		prefixListId *string
+	)
+
+	if e.in.Mft != nil {
```
same here

efekarakus replied:
👍 can we just return `nil` for now so that it's guaranteed that we don't accidentally include the `cdn` partial to the environment template?

iamhopaul123 commented on internal/pkg/manifest/env.go:
```diff
@@ -69,6 +71,73 @@ type environmentVPCConfig struct {
+// CDNEnabled returns whether a CDN configuration has been enabled in the environment manifest.
+func (e *environmentCDNConfig) CDNEnabled() bool {
+	if e.EnableCDN != nil {
+		if *e.EnableCDN {
+			return true
+		}
```
```suggestion
		return *e.EnableCDN
```

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/env.go:
```diff
@@ -296,6 +297,33 @@ func (e *BootstrapEnvStackConfig) ToEnv(stack *cloudformation.Stack) (*config.En
+	)
+
+	if e.in.Mft != nil {
+		// CDN Config specified by AdvancedCDNConfig
+		if !e.in.Mft.CDN.CDNConfig.IsEmpty() {
+			enabled = true
+			prefixListId = e.in.PrefixListID
+		}
+
+		// CDN Config specified by basic boolean enabler
+		if e.in.Mft.CDN.EnableCDN != nil {
+			if *e.in.Mft.CDN.EnableCDN {
+				enabled = true
+			}
+		}
```
I know your current implementation is the optimal one (with fewest if-else). However, since it doesn't really take much longer (could be ignored) and might make it more readable and less repetitive code we could do
```suggestion
                enable = e.in.Mft.CDN.CDNEnabled()
                if !e.in.Mft.CDN.CDNConfig.IsEmpty() {
                    prefixListId = e.in.PrefixListID
                }
```
since stack pkg doesn't really care how mft package is implemented

iamhopaul123 commented on internal/pkg/template/env.go:
```diff
@@ -59,6 +59,7 @@ func LeastVersionForFeature(feature string) string {
 var (
 	// Template names under "environment/partials/".
 	envCFSubTemplateNames = []string{
+		"cdn-resources",
```
oh nice actually this is a good finding

iamhopaul123 commented on internal/pkg/template/templates/environment/cf.yml:
```diff
@@ -56,6 +56,9 @@ Conditions:
   HasAliases:
     !Not [!Equals [ !Ref Aliases, "" ]]
 Resources:
+{{- if .CDNConfig.EnableCDN}}
```
Can we make CFN changes a separate PR 🙏 There are a lot to digest and by making another PRs we could potentially reduce the back-and-forth

iamhopaul123 commented on internal/pkg/manifest/env.go:
```diff
@@ -69,6 +71,73 @@ type environmentVPCConfig struct {
+// UnmarshalYAML overrides the default YAML unmarshaling logic for the environmentCDNConfig
+// struct, allowing it to perform more complex unmarshaling behavior.
+// This method implements the yaml.Unmarshaler (v3) interface.
+func (e *environmentCDNConfig) UnmarshalYAML(value *yaml.Node) error {
```
We might need to update the transformer file since this is a union type?

efekarakus commented on internal/pkg/template/templates/environment/partials/cdn-resources.yml:
```diff
@@ -0,0 +1,24 @@
+cloudfrontdistribution:
```
nit: to be consistent with our other logical IDs
```suggestion
CloudFrontDistribution:
```

efekarakus commented on internal/pkg/template/templates/environment/cf.yml:
```diff
@@ -100,6 +103,14 @@ Resources:
       GroupDescription: Access to the public facing load balancer
+       # Change load balancer to only accept ingress from CloudFront when prefix list specified
+{{- if .CDNConfig.PrefixListID}}
+      SecurityGroupIngress:
```
Should this be a list? https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-security-group.html#cfn-ec2-securitygroup-securitygroupingress
```suggestion
      SecurityGroupIngress:
        - SourcePrefixListId:
```

CaptainCarpensir replied:
For this and the below suggestion, it's relevant to the prefix list, so I'll address it in the next PR, and am removing these changes from the current one to reduce the scope. See comment below that as well.

efekarakus commented on internal/pkg/template/templates/environment/cf.yml:
```diff
@@ -100,6 +103,14 @@ Resources:
+        SourcePrefixListId: {{.CDNConfig.PrefixListID}}
+        FromPort: 0
+        IpProtocol: tcp
+        ToPort: 0
```
Is there a way of narrowing this down to be more specific for example can we do:
```yaml
        FromPort: 443
        IpProtocol: tcp
        ToPort: 443 # and another rule ToPort: 80
```

CaptainCarpensir replied:
With a max weight of 60, the cloudfront prefix list contains 55 of this, and so we can only have one security group for all of cloudfront.
https://docs.aws.amazon.com/vpc/latest/userguide/working-with-aws-managed-prefix-lists.html#aws-managed-prefix-list-weights
The only way I can think of to limit the port range is by specifying a max of 443 and a min of 80, but that seems pointless, as we then allow each port between then. I'd rather just allow cloudfront access to all ports in that case.

efekarakus commented on internal/pkg/template/templates/environment/partials/cdn-resources.yml:
```diff
@@ -0,0 +1,24 @@
+    DistributionConfig:
+      DefaultCacheBehavior:
+        AllowedMethods: ["GET", "HEAD", "OPTIONS", "PUT", "PATCH", "POST", "DELETE"]
+        CachePolicyId: 4135ea2d-6df8-44a3-9df3-4b5a84be39ad
```
😮 How is this value determined?

CaptainCarpensir replied:
The cache policy id is just a global id given by AWS, so I'm fairly certain can be hardcoded in this code.
[Here](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/using-managed-cache-policies.html) is more about this.

CaptainCarpensir replied:
The cache policy id is just a global id given by AWS, so I'm fairly certain can be hardcoded in this code.
[Here](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/using-managed-cache-policies.html) is more about this.

efekarakus replied:
Nice! can you add this as a comment next to the value?
```suggestion
        CachePolicyId: 4135ea2d-6df8-44a3-9df3-4b5a84be39ad # See https://go.aws/3bJid3k
```

efekarakus commented on internal/pkg/manifest/env.go:
```diff
@@ -69,6 +71,73 @@ type environmentVPCConfig struct {
 
+type environmentCDNConfig struct {
+	EnableCDN *bool
+	CDNConfig AdvancedCDNConfig // mutually exclusive with EnableCDN
```
nit:
```suggestion
	Enabled *bool
	Advanced AdvancedCDNConfig // mutually exclusive with Enabled
```

efekarakus commented on internal/pkg/manifest/env.go:
```diff
@@ -69,6 +71,73 @@ type environmentVPCConfig struct {
+
+// AdvancedCDNConfig represents an advanced configuration for a Content Delivery Network.
+type AdvancedCDNConfig struct {
+	PrefixListIngress *bool `yaml:"public_ingress_allowed"`
```
I wonder if this field will be clear enough to customers, what do you think of flipping the statement so that the field is by default false?
```suggestion
	LimitIngressToCDN *bool `yaml:"limit_origin_ingress"` 
```

CaptainCarpensir replied:
Will keep in mind as part of why I'm now splitting this PR is so that I can move this field to the environment ALB (in environmentHTTPConfig) so as to more accurately reflect where the security group resides. And yes, personally I think it's better to have public ingress disabled by default, so I'll change this for the next PR after splitting this one.

efekarakus commented on internal/pkg/manifest/env.go:
```diff
@@ -69,6 +71,73 @@ type environmentVPCConfig struct {
+			return true
+		}
+	}
+	return false
```
```suggestion
	return aws.BoolValue(cfg.Enabled)
```

efekarakus commented on internal/pkg/manifest/env.go:
```diff
@@ -69,6 +71,73 @@ type environmentVPCConfig struct {
+}
+
+// IsEmpty returns whether AdvancedCDNConfig is empty.
+func (a *AdvancedCDNConfig) IsEmpty() bool {
```
nit: can we rename all Config receivers to
```suggestion
func (cfg *AdvancedCDNConfig) IsEmpty() bool {
```

efekarakus commented on internal/pkg/manifest/env_test.go:
```diff
@@ -617,3 +656,79 @@ func TestEnvironmentObservability_IsEmpty(t *testing.T) {
+	}
+}
+
+func TestEnvironmentCDNConfig_CDNEnabled(t *testing.T) {
```
nice test cases!

Commit:
refactor pr scope

add transformer, remove prefix-list logic, address efekarakus and iamhopaul's feedback

CaptainCarpensir committed:
Merge branch 'mainline' into env-manifest-cdn

Commit:
fix missed merge overwrite conflict

iamhopaul123 commented on internal/pkg/manifest/env_test.go:
```diff
@@ -801,3 +819,65 @@ func TestEnvironmentObservability_IsEmpty(t *testing.T) {
+			},
+			wanted: true,
+		},
+		// "enabled via struct": {
+		// 	in: environmentCDNConfig{
+		// 		CDNConfig: AdvancedCDNConfig{},
+		// 	},
+		// 	wanted: true,
+		// },
```
🤔  can we get rid of unused code?

CaptainCarpensir replied:
Yeah, was there to remind myself to add back that test case when CDNConfig has fields to specify, but as there are other things to change as well like `IsEmpty()` it's not needed.

CaptainCarpensir replied:
There's similarly commented code in `transform_test.go` that I can get rid of too

iamhopaul123 commented on internal/pkg/manifest/env.go:
```diff
@@ -98,6 +100,57 @@ type environmentVPCConfig struct {
+}
+
+// AdvancedCDNConfig represents an advanced configuration for a Content Delivery Network.
+type AdvancedCDNConfig struct{}
```
Does this have to be public?

iamhopaul123 commented on internal/pkg/template/env.go:
```diff
@@ -103,10 +104,16 @@ type EnvOpts struct {
 }
 
+// CDNConfig represents a Content Delivery Network deployed by CloudFront.
+type CDNConfig struct {
+}
```
very nit
```suggestion
type CDNConfig struct {}
```

iamhopaul123 commented on internal/pkg/template/templates/environment/partials/cdn-resources.yml:
```diff
@@ -0,0 +1,24 @@
+      IPV6Enabled: true
+      Origins:
+        - CustomOriginConfig:
+            OriginProtocolPolicy: http-only # Will need to change with TLS termination and Alias usage
```
Why is this `http-only`? Does that mean CF will always terminate TLS for users?

Commit:
address iamhopaul123's feedback

iamhopaul123 commented on internal/pkg/manifest/env.go:
```diff
@@ -98,6 +100,57 @@ type environmentVPCConfig struct {
+	return cfg.Enabled == nil && cfg.CDNConfig.IsEmpty()
+}
+
+// IsEmpty returns whether AdvancedCDNConfig is empty.
```
Maybe specify this is a no-op now

iamhopaul123 commented on internal/pkg/manifest/env.go:
```diff
@@ -98,6 +100,57 @@ type environmentVPCConfig struct {
+// This method implements the yaml.Unmarshaler (v3) interface.
+func (cfg *environmentCDNConfig) UnmarshalYAML(value *yaml.Node) error {
+	if err := value.Decode(&cfg.CDNConfig); err != nil {
+		switch err.(type) {
+		case *yaml.TypeError:
+			break
+		default:
+			return err
+		}
```
```suggestion
		var yamlTypeErr *yaml.TypeError
		if !errors.As(err, &yamlTypeErr) {
			return err
		}
```

efekarakus commented on internal/pkg/manifest/transform_test.go:
```diff
@@ -1124,3 +1124,55 @@ func TestSecretTransformer_Transformer(t *testing.T) {
+		override func(cfg *environmentCDNConfig)
+		wanted   func(cfg *environmentCDNConfig)
+	}{
+		// "enabled set to nil if cdnconfig is not empty": {
```
can we delete the unused comments? 🥺

efekarakus reviewed (commented):
Looks great! 👍 to Penghao's feedback and 2 small requests, and then we can :shipit:

Commit:
address efekarakus and iamhopaul123's feedback

Commit:
nit

mergify[bot] committed:
Merge branch 'mainline' into env-manifest-cdn
